### ISSUE TITLE
Help: I ​​trained the text-to-pose model myself, but the prediction results are always poor

### ISSUE THREAD

**Issue by SWHL** [clement-bonnet/text-to-pose#4]
Thanks for your great work. Recently I am trying myself text-to-pose model based on the repo. I used 500,000 data sets labeled with DWPose and JoyCaption. During the training process, the loss has been decreasing, but the prediction results for the test set have been very poor. I would like to ask you what might be the reason?

The loss:

![image](https://github.com/user-attachments/assets/8d869c1a-bf63-4cc9-ba34-f678896627c8)

The results of the eval prompt:

![image](https://github.com/user-attachments/assets/799550ec-5d83-46c8-894a-abf4c3dccace)

**Comment by clement-bonnet** [comment]
Your loss values seem a bit off to me. I don't recall such high spikes, so there could be something going on with that. Otherwise, I do recall that you need quite a bit of training before seeing good results. Quite quickly, the body key points get predicted correctly but faces and hands take much longer. So, it could just be that you need to train for longer.
Hope this helps! Alternatively, you can drop some code/training snippets for me to have a look.

**Comment by SWHL** [comment]
Thank you for your response. I am writing training code based on your open-source code. I have two questions:
1. In the training set, are the coordinates of pose keypoints relative to the image, or are they normalized to the bounding boxes of the human body?
2. Are the coordinates of human body bouncing boxes and other human pose keypoints trained together? Or is it to train these two parts in stages?

**Comment by clement-bonnet** [comment]
1. In the data processing stage, we rescale coordinates to be relative to their bounding box. But you could change that if you wish.
2. A pose sequence is composed of the following: [bounding box 1, ..., bounding box N, pose 1, ..., pose N], everything is trained together, in one stage.

**Comment by SWHL** [comment]
Thanks, we want to reproduce text-to-pose great work, and train our model.

May I ask if it is convenient for you to provide the source code for training here? If it's inconvenient,  I don't think it's a problem. 

Thank you again.

My email is liekkaskono@163.com

**Comment by clement-bonnet** [comment]
It's not that I don't want but unfortunately, I do not have the training script anymore. Yet I can tell you it was a very simple script that was processing the data and doing backprop on this loss: https://github.com/clement-bonnet/text-to-pose/blob/main/t2p/model.py#L135

**Comment by SWHL** [comment]
Thanks for your response.  
I have written the code of processing the data and the train code, the part of `train.py` is like this:

<details>
 <summary>Click to expand</summary>

```python
import torch

from t2p.t2p_sample import T2PSample


def extract_embedding(prompt: str):
    with torch.no_grad():
        input_ids = clip_processor(
            text=prompt, return_tensors="pt", padding="max_length", truncation=True
        )["input_ids"]

        text_embeddings = (
            model.clip_text_model(input_ids.to(device), output_hidden_states=True)
            .hidden_states[-2]
            .squeeze()
        )
    return text_embeddings


def get_init_t2psample(
    num_poses: int,
    max_num_poses: int,
    image_ratio,
    device,
    dtype,
    bboxes=None,
    pose_mask=None,
):
    if bboxes is None:
        bboxes = torch.full(
            (num_poses, max_num_poses, 2, 2),
            -1.0,
            device=device,
            dtype=dtype,
        )

    if pose_mask is None:
        pose_mask = torch.zeros(
            num_poses, max_num_poses, device=device, dtype=torch.bool
        )

    poses = T2PSample(
        bboxes=bboxes,
        bodies=torch.full(
            (num_poses, max_num_poses, 18, 2),
            -1.0,
            device=device,
            dtype=dtype,
        ),
        faces=torch.full(
            (num_poses, max_num_poses, 68, 2),
            -1.0,
            device=device,
            dtype=dtype,
        ),
        hands=torch.full(
            (num_poses, max_num_poses, 42, 2),
            -1.0,
            device=device,
            dtype=dtype,
        ),
        pose_mask=pose_mask,
        image_ratio=image_ratio,
        input_ids=None,
        uuid=None,
    )
    return poses


def train(model, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, gt_data in enumerate(train_loader):
        text_embeddings = torch.stack(
            [extract_embedding(v) for v in gt_data.prompt], dim=0
        )
        optimizer.zero_grad()

        cur_bt = gt_data.bodies.shape[0]
        input_data = get_init_t2psample(
            cur_bt,
            max_num_poses=5,
            image_ratio=gt_data.image_ratio,
            pose_mask=gt_data.pose_mask,
            device=device,
            dtype=dtype,
        )
        logits = model(input_data, text_embeddings)
        loss = model.loss(logits, gt_data)
    
        loss["loss"].backward()
        optimizer.step()

```
</details>

And I wrote the data preprocessing part as follows:

<details>
 <summary>Click to expand</summary>

```python

class TextToPoseMultiPersonDataset(Dataset):
    def __init__(
        self,
        jsonl_path: Union[str, Path, List[str]],
        dwpose_body_th: float = 0.3,
        max_num_poses: int = 5,
    ):
        self.datas = read_jsonl(jsonl_path)
        self.dwpose_body_th = dwpose_body_th
        self.max_num_poses = max_num_poses
        self.device = torch.device("cpu")
        self.dtype = torch.float32
        self.fix_img_shape = (1024, 1024)  # w, h

    def __getitem__(self, idx):
        data = self.datas[idx]
        img_path = data["img_path"]
        # print(img_path)

        pose_info = data["pose_data"]
        self.img_w, self.img_h = data.get("width"), data.get("height")

        self.num_poses = len(pose_info["bboxes"])
        prompt = self.get_caption(data)

        if self.num_poses > self.max_num_poses or len(prompt) <= 0:
            return self.__getitem__(random.randint(0, len(self.datas) - 1))

        bboxes = self.get_bboxes(pose_info["bboxes"])

        if len(bboxes) <= 0:
            return self.__getitem__(random.randint(0, len(self.datas) - 1))

        self.bbox_x0, self.bbox_x1 = copy.deepcopy(bboxes[:, :, 0]).T
        self.bbox_y0, self.bbox_y1 = copy.deepcopy(bboxes[:, :, 1]).T
        self.bboxes_w = torch.abs(self.bbox_x1 - self.bbox_x0)
        self.bboxes_h = torch.abs(self.bbox_y1 - self.bbox_y0)

        bboxes[..., 0] /= self.img_w
        bboxes[..., 1] /= self.img_h

        new_bboxes = torch.full(
            (self.max_num_poses, 2, 2), -1, dtype=bboxes.dtype, device=bboxes.device
        )
        new_bboxes[: self.num_poses, :, :] = bboxes

        bodies_points = self.get_bodies_points(pose_info["bodies"])

        faces_points = self.get_faces_points(
            pose_info["faces"], pose_info["faces_score"]
        )

        hands_points = self.get_hands_points(
            pose_info["hands"], pose_info["hands_score"]
        )

        pose_mask = self.get_pose_mask()
        image_ratio = self.get_image_ratio()

        poses = T2PSample(
            bboxes=new_bboxes,
            bodies=bodies_points,
            faces=faces_points,
            hands=hands_points,
            pose_mask=pose_mask,
            image_ratio=image_ratio,
            input_ids=None,
            uuid=None,
            prompt=prompt,
        )

        return poses

    def __len__(self):
        return len(self.datas)

    @staticmethod
    def get_caption(data):
        if "florence_caption" in data:
            caption = data.get("florence_caption").strip()
            return caption

        if "caption" in data:
            caption = data.get("caption").strip()
        return caption

    def get_bboxes(self, bboxes):
        bboxes = torch.Tensor(bboxes)
        bboxes = torch.where(bboxes < 0, 0, bboxes)
        bboxes = bboxes.reshape(-1, 2, 2)
        return bboxes

    def get_bodies_points(self, bodies_info) -> torch.Tensor:
        subsets = bodies_info["subset"]
        scores = bodies_info["score"]

        candidates = bodies_info["candidate"]
        candidates = [candidates[i : i + 18] for i in range(0, len(candidates), 18)]

        new_points = []
        for i, (candidate_one, subset_one, score_one) in enumerate(
            zip(candidates, subsets, scores)
        ):
            for candidate, subset, score in zip(candidate_one, subset_one, score_one):
                if score < self.dwpose_body_th or subset == -1:
                    new_points.append([torch.Tensor([-1.0]), torch.Tensor([-1.0])])
                    continue

                raw_x = candidate[0] * self.img_w
                raw_y = candidate[1] * self.img_h
                new_candidate = [
                    (raw_x - self.bbox_x0[i]) / self.bboxes_w[i],
                    (raw_y - self.bbox_y0[i]) / self.bboxes_h[i],
                ]
                new_points.append(new_candidate)
        new_points = torch.Tensor(new_points).reshape(self.num_poses, 18, 2)

        new_bodies_points = torch.full(
            (self.max_num_poses, 18, 2),
            -1,
            dtype=self.dtype,
            device=self.device,
        )
        new_bodies_points[: self.num_poses, :, :] = new_points
        return new_bodies_points

    def get_faces_points(self, faces, face_scores):
        total_faces = []
        for i, (one_face, one_score) in enumerate(zip(faces, face_scores)):
            new_faces = []
            for face, score in zip(one_face, one_score):
                if score < self.dwpose_body_th:
                    new_faces.append([torch.Tensor([-1.0]), torch.Tensor([-1.0])])
                    continue

                raw_x = face[0] * self.img_w
                raw_y = face[1] * self.img_h
                new_face = [
                    (raw_x - self.bbox_x0[i]) / self.bboxes_w[i],
                    (raw_y - self.bbox_y0[i]) / self.bboxes_h[i],
                ]
                new_faces.append(new_face)
            total_faces.append(new_faces)
        total_faces = torch.Tensor(total_faces).reshape(self.num_poses, 68, 2)

        new_faces_points = torch.full(
            (self.max_num_poses, 68, 2),
            -1,
            dtype=self.dtype,
            device=self.device,
        )
        new_faces_points[: self.num_poses, :, :] = total_faces
        return new_faces_points

    def get_hands_points(self, hands, hand_scores) -> torch.Tensor:
        group_hands = [
            [i for i in range(j, len(hands), self.num_poses)]
            for j in range(self.num_poses)
        ]
        bboxes_map_hands = {i: group_hands[i] for i in range(self.num_poses)}

        bboxes_idx = None
        total_hands_points = {}
        for i, (hand_points, hand_score) in enumerate(zip(hands, hand_scores)):
            for k, v in bboxes_map_hands.items():
                if i in v:
                    bboxes_idx = k

            if bboxes_idx is None:
                raise ValueError("Hands Data Error")

            one_hands_points = self.get_one_hand_points(
                hand_points, hand_score, bboxes_idx
            )
            total_hands_points.setdefault(bboxes_idx, []).append(one_hands_points)
        hands_points = torch.Tensor(list(total_hands_points.values())).reshape(
            self.num_poses, 42, 2
        )

        new_hands_points = torch.full(
            (self.max_num_poses, 42, 2),
            -1,
            dtype=self.dtype,
            device=self.device,
        )
        new_hands_points[: self.num_poses, :, :] = hands_points
        return new_hands_points

    def get_one_hand_points(self, hands, hand_scores, i):
        new_hands = []
        for hand, score in zip(hands, hand_scores):
            if score < self.dwpose_body_th:
                new_hands.append([torch.Tensor([-1.0]), torch.Tensor([-1.0])])
                continue

            raw_x = hand[0] * self.img_w
            raw_y = hand[1] * self.img_h
            hand = [
                (raw_x - self.bbox_x0[i]) / self.bboxes_w[i],
                (raw_y - self.bbox_y0[i]) / self.bboxes_h[i],
            ]
            new_hands.append(hand)
        return new_hands

    def get_pose_mask(self):
        pose_mask = torch.ones(self.num_poses).bool()
        new_pose_mask = torch.zeros(
            self.max_num_poses, device=self.device, dtype=torch.bool
        )
        new_pose_mask[: self.num_poses] = pose_mask
        return new_pose_mask

    def get_image_ratio(self):
        image_ratio = self.img_w / self.img_h
        return torch.tensor(image_ratio)

```
</details>

If it's convenient for you, could you help me take a look at the code? I really can't find where the problem is. Thank you.

Or if you don't feel comfortable discussing it here, you can also discuss it on Twitter, I'm all for that here.

**Comment by clement-bonnet** [comment]
Do you call `train(model, train_loader, optimizer, epoch)` at each gradient step? I don't understand why you would not call backward after each batch from the train_loader.

**Comment by SWHL** [comment]
It's a typo, I had correct it. I do call backward after each batch from the train_loader.

**Comment by RYG81** [comment]
@SWHL did u managed to get things working?

**Comment by SWHL** [comment]
@RYG81 After trying all the methods I know, I still haven't been able to converge well. I gave up.

**Comment by Genghao-025** [comment]
> Thanks for your great work. Recently I am trying myself text-to-pose model based on the repo. I used 500,000 data sets labeled with DWPose and JoyCaption. During the training process, the loss has been decreasing, but the prediction results for the test set have been very poor. I would like to ask you what might be the reason?
> 
> The loss:
> 
> ![image](https://github.com/user-attachments/assets/8d869c1a-bf63-4cc9-ba34-f678896627c8)
> 
> The results of the eval prompt:
> 
> ![image](https://github.com/user-attachments/assets/799550ec-5d83-46c8-894a-abf4c3dccace)
May I ask if you are using the "gmm" loss in the code here? What is the batch_size set to? I only trained the generation of 18-point poses for a single person, and all the predicted coordinates are negative. I don't know the reason.

**Comment by SWHL** [comment]
@Genghao-025 I used the same loss with the repo. As the following is: 

https://github.com/clement-bonnet/text-to-pose/blob/bd6111b5469352af3ccd863ce43cdb687185b3e3/t2p/model.py#L183

 My training results have always been poor, and I still haven't figured out what went wrong.
